samples/Ralltiir/Ralltiir.Server/Controllers/AuthorizationController.cs

```diff
@@ -0,0 +1,343 @@
+                case ConsentTypes.Explicit when authorizations.Any() && !request.HasPrompt(Prompts.Consent):
+                    var principal = await _signInManager.CreateUserPrincipalAsync(user);
+
+                    // Note: in this sample, the granted scopes match the requested scope
+                    // but you may want to allow the user to uncheck specific scopes.
+                    // For that, simply restrict the list of scopes before calling SetScopes.
+                    principal.SetScopes(request.GetScopes());
+                    principal.SetResources(await _scopeManager.ListResourcesAsync(principal.GetScopes()).ToListAsync());
+
+                    // Automatically create a permanent authorization to avoid requiring explicit consent
+                    // for future authorization or token requests containing the same scopes.
+                    var authorization = authorizations.LastOrDefault();
+                    if (authorization == null)
+                    {
+                        authorization = await _authorizationManager.CreateAsync(
+                            principal: principal,
+                            subject  : await _userManager.GetUserIdAsync(user),
+                            client   : await _applicationManager.GetIdAsync(application),
+                            type     : AuthorizationTypes.Permanent,
+                            scopes   : principal.GetScopes());
+                    }
+
+                    principal.SetAuthorizationId(await _authorizationManager.GetIdAsync(authorization));
+
+                    foreach (var claim in principal.Claims)
+                    {
+                        claim.SetDestinations(GetDestinations(claim, principal));
+                    }
+
+                    return SignIn(principal, OpenIddictServerAspNetCoreDefaults.AuthenticationScheme);
+
+                // At this point, no authorization was found in the database and an error must be returned
+                // because the client application specified prompt=none in the authorization request.
+                case ConsentTypes.Explicit:
+                case ConsentTypes.Systematic:
+                    return Forbid(
+                        authenticationSchemes: OpenIddictServerAspNetCoreDefaults.AuthenticationScheme,
+                        properties: new AuthenticationProperties(new Dictionary<string, string>
+                        {
+                            [OpenIddictServerAspNetCoreConstants.Properties.Error] = Errors.ConsentRequired,
+                            [OpenIddictServerAspNetCoreConstants.Properties.ErrorDescription] =
+                                "Interactive user consent is required."
+                        }));
+
+                // In every other case, return an error since we only support promptless responses
+                default: throw new NotImplementedException("The specified grant type is not implemented.");
```

Comment:
I suspect this is one thing you'll absolutely need to change. At this point, it's the "render a consent view" case, so you need to somehow flow the consent details from this controller to your SPA frontend, maybe by using `ViewData` to store those details and returning a `View` that loads the SPA assets and points to the `authorize.tsx` page. Most likely, it's a scenario that will require reloading the SPA as the request needs to reach the ASP.NET Core server so that OpenIddict can validate the authorization request parameters.